New features for Linux

robxnano opened the pull request:
Hello again, I've been working on some of the most requested features for the Linux version, as well as a bit of maintenance to fix some warnings caused by deprecated functions.

- Drag and drop for videos
- Extra diagnostic information in the log
- Pause encoding when running on battery power (https://github.com/HandBrake/HandBrake/issues/2109)
- XML chapter import and export (fixes https://github.com/HandBrake/HandBrake/issues/3179)
- Native file chooser (fixes https://github.com/HandBrake/HandBrake/issues/238)
- Fix for symbolic icons on older GTK version (fixes https://github.com/HandBrake/HandBrake/issues/4884)

The native file chooser also allows opening videos from some non-native file systems such as MTP and network drives. Previously these were not accessible without copying to local storage. I haven't yet managed to get them to work properly with Flatpak builds though.

I have some other changes I'd like to make, but they will need more discussion. These are the relatively straightforward ones.

robxnano committed:
LinGUI: Bump GTK 3 minimum version to 3.22

Should be available in all currently supported distributions.
Also removed conditionals that are no longer needed.

robxnano committed:
LinGUI: Use GtkFileChooserNative

When xdg-desktop-portal is available, a dialog native to
the desktop environment is used instead of the GTK one.
Fixes #4700

robxnano committed:
LinGUI: Bump GTK 4 minimum version to 4.4

GTK 4 support is currently based on a beta version and doesn't work
due to deprecations in the release versions. Further changes are
needed before it will build successfully.

robxnano committed:
LinGUI: Refactor Add Multiple dialog

Moves all the code for the Add Multiple dialog from
queuehandler.c to title-add.c and renames functions and
actions to match. This will help make the code easier to
maintain and allow for further improvements.

robxnano committed:
LinGUI: Add menu option to Add All items

Adds all titles to the queue immediately. If the destination
filenames are not unique, opens the Add Multiple dialog instead.
Also stops adding items if an error occurs instead of continuing
and creating repeated error messages (Fixes #4438).

robxnano committed:
LinGUI: Drag and drop support for videos

Drag a video or folder onto the HandBrake window and it will be
scanned as if it had been opened through the file dialog. Works
with most file managers (Nautilus, Dolphin, Nemo, Thunar etc.)
Closes #969

robxnano committed:
LinGUI: Rearrange Filter tab to match other platforms

robxnano committed:
LinGUI: Add XML chapter import and export

- Imports chapter titles only
- Exports full chapter details including time stamps
- Supports the Matroska XML chapter format
- Reads chapters from the default EditionEntry
- Uses the first ChapterString found for each chapter

robxnano committed:
LinGUI: Fix symbolic icons on old GTK versions

robxnano committed:
LinGUI: Fix function prototype warnings

Fixes most warnings caused by old-style function declarations
and non-static functions with no header declarations.
Warnings were found by compiling in GCC 12 with the flags:
-Wmissing-declarations
-Wmissing-parameter-type
-Wmissing-prototypes
-Wold-style-declaration
-Wold-style-definition
-Wstrict-prototypes

robxnano committed:
LinGUI: Use structured logging

Replaces the custom logging functions with standard GLib ones,
using the __func__ macro introduced in C99 for function names.
Instead of using the --debug argument, debug messages are shown if
the environment variable G_MESSAGES_DEBUG is set to 'ghb' or 'all'.

robxnano committed:
Linux: Add more diagnostic info to log

Prints the current operating system, kernel version and data
directories to the activity log.

robxnano committed:
LinGUI: Use native paths for dropped files

Try to get a native path instead of a URI for files added via
drag and drop so that libhb can access them through GVFS.

robxnano committed:
Fix warning if dialog is created with no OK button

robxnano committed:
Linux: Add CPU info to activity log

- Remove temp dir info as it is often inaccurate

robxnano committed:
LinGUI: Automatic pause to save battery power

- Listens on DBus for battery status events
- Pauses encoding if unplugged or entering power save mode
- Resumes when power is restored or exiting power save mode
- Encodes can still be started while on battery power
- Doesn't resume when plugged in if paused manually
- Preferences option for pause on unplug

robxnano committed:
LinGUI: Run autoupdate to fix configure.ac warnings

robxnano committed:
LinGUI: New auto-naming options

{codec} - The codec of the encoded video (such as H.264)
{bit-depth} - The bit depth for the encoded video as an integer
{width} - The storage width of the encoded video
{height} - The storage height of the encoded video
{modification-date} - The modification date of the source file
{modification-time} - The modification time of the source file

robxnano committed:
LinGUI: Update translation template

- Rename metainfo template so that gettext recognizes it

robxnano committed:
LinGUI: Removed leftover version guards

sr55 commented:
Cool. Thanks for the updates. I'll try make some time this weekend to test this lot :)

sr55 commented:
![image](https://user-images.githubusercontent.com/628593/218164815-e34abc3d-2dbc-46ac-ae86-ad9543520bc5.png)

The wording there is a bit awkward.  Maybe:

"Pause Encode when switching to battery"

The WinUI actually says "Pause any running jobs when the battery gets low." and has a "Low battery level:" field too.  Doesn't really need to be that granular in my view.

sr55 commented:
Not sure if this is deliberate, or something to do with the fact that I'm running it under WSL (on windows) but is there any reason the Main UI grey's out when the log window opens:

![image](https://user-images.githubusercontent.com/628593/218167114-2763ca73-0f45-41f7-bfe6-fc9fcf3c5645.png)


You can still mostly interact with it. Seems a bit odd to behave like this as you'd maybe want the log window off to the side for monitoring

sr55 commented:
Can't test drag/drop under WSL but the rest seems to be good.

robxnano commented:
> Not sure if this is deliberate, or something to do with the fact that I'm running it under WSL (on windows) but is there any reason the Main UI grey's out when the log window opens:
> 
> ![image](https://user-images.githubusercontent.com/628593/218167114-2763ca73-0f45-41f7-bfe6-fc9fcf3c5645.png)
> 
> You can still mostly interact with it. Seems a bit odd to behave like this as you'd maybe want the log window off to the side for monitoring

I think this is just how the default theme looks when the window is not focused. It may look a little odd on Windows but on Linux it will look the same as all other GTK applications.

robxnano commented:
> The wording there is a bit awkward. Maybe:
> 
> "Pause Encode when switching to battery"
> 
> The WinUI actually says "Pause any running jobs when the battery gets low." and has a "Low battery level:" field too. Doesn't really need to be that granular in my view.

I'll change the wording. Do you think there should be another checkbox for pausing when the battery is low? Right now it's using the Power Saver profile as a proxy for low battery, but I could change it to use the actual battery level.
Also I was thinking of having something to explain why it was paused, either by changing the status message to say something like "Encoding paused as computer was unplugged", or by showing a notification.

bradleysepos commented:
Something like this might make sense:

Pause encode jobs when:

☑️ Power Saver is enabled
☑️ Battery is low
☑️ Switching to battery power

Feels like switching to battery power should be off by default?

robxnano committed:
LinGUI: Miscellaneous fixes

- Free unneeded filename
- Fix XML chapter export crash
- Replace unneeded printf calls with g_debug
- Update deprecated margin-left property

robxnano committed:
LinGUI: Access D-Bus asynchronously

robxnano committed:
LinGUI: Option to pause on entering power save

robxnano committed:
LinGUI: Pause encoding when battery is low

robxnano committed:
LinGUI: Tweak power save options

Although there's no option for it in the UI, adding LowBatteryLevel
to preferences.json will set the level at which encoding is paused.

robxnano committed:
LinGUI: Update translation files

robxnano commented:
Added the full set of options to the preferences window. By default, only the Power Save option is activated. The bottom two options will only be shown if a battery is present. The low battery level is currently 15%, although this can be changed by editing preferences.json.

![Screenshot from 2023-02-18 18-43-35](https://user-images.githubusercontent.com/89391914/219883332-9f972e32-2362-4d7c-86fc-de92f054119d.png)

bradleysepos commented:
Nice, looks good and very useful. Most minor nit: Pause encoding *when* in Power Saver mode?

robxnano commented:
Yeah that sounds better.

sr55 commented:
Ok, Lets push this and get some user feedback :)

Thanks for your efforts!

Nomis101 commented:
Thanks for your effort robxnano. I would like to have a localizable related questions. We do all our localizations via Transifex. Which means that all locale files will get a regular update with ne localized strings. There is now a new string "`Add A_ll`". What does this stand for? Is this a "Add all" with one of those strange linux specific underscores? Or is this A and a roman 2?

robxnano commented:
Hi @Nomis101. Yes, the correct string is just Add All. The underscores are used in menus to denote an accelerator key, which means that the 'l' is displayed with an underline to show that it can be activated with the L key. By the way, does this cause any problems for translators? I can imagine how it would be hard to know which key to use as the accelerator without seeing the whole list of menu items. But it's not really a problem if more than one item uses the same letter, as in that case the key just cycles through all of them.

Let me know if there's anything else I can do to make your localisation work easier. I'd be happy to help if there's something that needs fixing or clarifying, and I can add translator's notes in the source if necessary.

robxnano commented:
Thanks @sr55. I forgot to mention, but I implemented some of the requested automatic output options, namely https://github.com/HandBrake/HandBrake/issues/3169 and https://github.com/HandBrake/HandBrake/issues/4778. https://github.com/HandBrake/HandBrake/issues/4532 is also not needed on Linux as all the date options use the ISO format already.

Nomis101 commented:
> Hi @Nomis101. Yes, the correct string is just Add All. The underscores are used in menus to denote an accelerator key, which means that the 'l' is displayed with an underline to show that it can be activated with the L key. By the way, does this cause any problems for translators? I can imagine how it would be hard to know which key to use as the accelerator without seeing the whole list of menu items. But it's not really a problem if more than one item uses the same letter, as in that case the key just cycles through all of them.
> 
> Let me know if there's anything else I can do to make your localisation work easier. I'd be happy to help if there's something that needs fixing or clarifying, and I can add translator's notes in the source if necessary.

OK, thanks for clarifying. Does this mean, if the underscore will be included in the translation, it will change the accelerator key? For example `_Quit` is translated to Corsican `_Esce` or `_Beenden` in German. Or should the underscore not be included in the translated string (if so, this will be tricky, because all locales are doing so)?

sr55 commented:
We used to have these keyboard accelerators in the Windows version too when it was WinForms. handy feature to have. I must have forgotten to re-add them when we moved to WPF many years ago. 

I guess it is awkward for translation as the _ needs to map to a keyboard key. So the optimal position can change for each language.

robxnano commented:
> OK, thanks for clarifying. Does this mean, if the underscore will be included in the translation, it will change the accelerator key? For example `_Quit` is translated to Corsican `_Esce` or `_Beenden` in German. Or should the underscore not be included in the translated string (if so, this will be tricky, because all locales are doing so)?

Yeah, the accelerator key will be whichever letter follows the underscore in the translated string.
